### ISSUE TITLE
Tick lag when autocrafting, profile snapshot attached

### ISSUE THREAD

**Issue by thetaphi** [AppliedEnergistics/Applied-Energistics-2#982]
**TL;DR:** I request autocrafting of lots of stuff, most of which can't be made at that moment. Server can't keep up. Might be NBT stuff, or a thread timing issue. Profile snapshot: 
https://github.com/thetaphi/tripping-octo-dangerzone/blob/master/FTB_ME4_thetaphi_fullAE2Setup.nps

### Software

AE2 rv2-beta-8

This is an ME^4 1.0.4 single player world, via FTB 1.4.5, with the following mods updated by me:
* Minechem to 5.0.5.385
* Fluxed Crystals to 0.0.5d
* AE2 to rv2-beta-8

I also installed Opis 1.2.3 for diagnostics. 

Java is build 1.8.0_40-b25.
Computer is running Windows 7 on an i7 2.3GHz and 8GB RAM.

### The crafting setup

I'm autocrafting up to octuple compressed cobble and cookies. [Two interfaces with crafting cards](http://i.imgur.com/AU5afqg.png) hold 9 each of the compression stages. [Storage buses](http://i.imgur.com/a5cIFWo.png) are on those interfaces. A [secondary AE system](http://i.imgur.com/jpTXNVQ.png) continuously crafts up to double compressed cobble and cookies, via crafting carded export buses. The interface above crafting requests stacks of cobble, cocoa and wheat.

Cocoa and wheat are autocrafted from iron and cobblestone by the primary AE system on demand, via pulverizing, Minechem breaking sand down to elements, fissioning silicon and iron, and recombining the products.

## What I see:

There's tick lag to the point where I have to break a cobblestone block thrice, and then it vanishes.

[F3 and Opis screenshot](http://i.imgur.com/ftTMswF.png)

Primary observations:

* The timing total does not add up
* Why does Opis tell me I have 300ms ping to localhost!?
* That "unspecified" chunk is rather large. BTW, that unspecified thing starts <5% upon world start, and then grows over 2-4 minutes to the point in the screenshot.
* Obv not visible in the screenshot, but "Used memory" goes from ~15% to >50% and back down about twice a second. GC carnage!

Drilling down in Opis did show every particular entity, chunk or whatnot timing in the microseconds range.

Console:
> [05:35:40] [Client thread/WARN]: Something's taking too long! 'root.gameRenderer' took aprox 115.313891 ms
> [05:35:40] [Client thread/WARN]: Something's taking too long! 'root' took aprox 122.195385 ms
> [05:35:49] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 43109ms behind, skipping 862 tick(s)
> [05:35:54] [Client thread/WARN]: Something's taking too long! 'root.gameRenderer' took aprox 100.271163 ms
> [05:35:54] [Client thread/WARN]: Something's taking too long! 'root' took aprox 108.339137 ms
> [05:35:55] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 5692ms behind, skipping 113 tick(s)
> [05:36:03] [Client thread/WARN]: Something's taking too long! 'root.gameRenderer.level' took aprox 103.066687 ms
> [05:36:03] [Client thread/WARN]: Something's taking too long! 'root.gameRenderer' took aprox 104.746678 ms
> [05:36:03] [Client thread/WARN]: Something's taking too long! 'root' took aprox 114.370151 ms

## What I did:

I started to notice tick lag a few days ago, and it is getting worse. Since my client frames are well above 20, and my CPU utilization never goes above 30% in total and about 60% per core, I decided to separate possible causes and set up a server with my world. The server on a spare linux machine exhibited *severe* tick lag right upon login, while on the client the warnings disappeared.

I poked around on my computer, and do not seem to have any process or service futzing around with localhost connections or the software. Malwarebytes found nothing suspicious.

Downgrading Fluxed Crystals and Minechem to the original versions did not help.

Then I looked at what was in the most utilized chunk, which was the centerpiece of my AE system.
Putting a rogue controller on a cable, thereby disabling the whole system, cleared up most of the tick lag. So, I restored everything to the full, lagging setup with the mod versions above, and created the profile snapshot.

Poking around in that snapshot showed [CraftingJob.handlePausing()](https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/rv2/src/main/java/appeng/crafting/CraftingJob.java#L325) (Message to the future: That link will deteriorate) as a hotspot, but drilling down the server thread call stack afterwards shows that most of the time goes into NBTTagCompound, via export bus crafting. And that's where I have no clue any more.

**Comment by thetaphi** [comment]
I ran a profile again with beta8 (was from beta3 originally) and edited OP accordingly.

**Comment by XFactHD** [comment]
In the log it says root.gameRenderer took 100.271163. It could be a problem with the rendering of some machines because minechem has some really complex modells. I got a rendering crash some time ago because of a cc screen that was hundreds of blocks away and could not be seen. Which graphics card do you use?

**Comment by thetaphi** [comment]
So, after I added the NBT suspicion as an afterthought, armed with that knowledge I found possibly related tickets like #670.

I'm currently redesigning my AE system to get around the NBT issue, splitting off a third branch and adding MOAR minechem machines in the process. So far it's a success, the tick lag is gone and the client/WARNs with it.

NVIDIA GeForce GT 630M, aka shitty laptop card, but it seems it's not the issue.

**Comment by yueh** [comment]
It might be related to that issue (in terms of being a side effect). But it is mostly about NBT being slow in general. As long as the items are inside the AE2 network, they will share the same NBT compound. But the moment they leave (or enter) the network they need to be turned into an item with a discrete NBT compound. If minechem is heavily using NBT this will of course affect it.

But the real issue here is, that the CellInventory converts the AE2 internal itemstack into a normal vanilla one. Just to access the stacksize, which is already known and the conversion is pretty useless.
We should clearly look into it. But it is not really related to the other issue, it just adds up.

**Comment by Slind14** [comment]
wow, that's a great solution. mind telling me in which build it is included so I can do some testing with the old setup?

**Comment by thatsIch** [comment]
seems it was rv2beta9

**Comment by Slind14** [comment]
alright, thanks for looking it up.
